docs/DevGuide/Brigand.md

```diff
@@ -0,0 +1,1827 @@
+
+
+\section function_docs Brigand functions
+
+\subsection Containers
+
+\par
+Brigand provides container classes with the sole purpose of wrapping other
+things.
+
+
+\subsubsection integral_constant
+
+\par
+Very similar to std::integral_constant, except that the `constexpr` specifiers
+on the member functions have been omitted.
+\snippet Test_TMPLDocumentation.cpp tmpl::integral_constant
+
+\par
+Brigand supplies type aliases for constants of some specific types:
+\snippet Test_TMPLDocumentation.cpp tmpl::integral_constant::abbreviations
+
+\par
+Most metafunctions that accept integral_constants will accept any type with a
+`value` static member variable.
+
+\par
+Because of the `type` type alias, integral_constants behave like lazy
+metafunctions returning themselves.  Most lazy metafunctions producing an
+integral_constant will actually inherit from their result, so `value` will be
+directly available without needing to go through the `type` alias.
+
+\remark
+Prefer std::integral_constant, except for the convenience wrapper
+`tmpl::size_t` or when necessary for type equality comparison.
+
+
+\subsubsection list
+
+\par
+An empty struct templated on a parameter pack, with no additional
+functionality.
+\snippet Test_TMPLDocumentation.cpp tmpl::list
+
+\par
+Most metafunctions that operate on lists will work on any struct template.
+
+
+\subsubsection map
+
+\par
+A collection of key-value \ref pair ""s with unique keys.  See the section on
+\ref map_operations "operations on maps" for details.
+\snippet Test_TMPLDocumentation.cpp tmpl::map
+
+\par
+The actual type of a map is implementation-defined, but it has the same
+template parameters as a call to `map` that would produce it.
+
+\warning
+Equivalent maps may have different types, depending on the order their keys are
+stored in internally.
+
+
+\subsubsection pair
+
+\par
+A pair of types, with easy access to each type in the pair.
+\snippet Test_TMPLDocumentation.cpp tmpl::pair
+
+
+\subsubsection set
+
+\par
+An unordered collection of distinct types.  Trying to create a `set`
+with duplicate entries is an error (but \ref set_insert "insert" ignores
+duplicate entries).
+\snippet Test_TMPLDocumentation.cpp tmpl::set
+
+\par
+The actual type of a set is implementation-defined, but it has the same
+template parameters as a call to `set` that would produce it.
+
+\warning
+Equivalent sets may have different types, depending on the order their elements
+are stored in internally.
+
+
+\subsubsection type_
+
+\par
+A struct templated on a single type `T` containing an alias `type` to `T`.
+\snippet Test_TMPLDocumentation.cpp tmpl::type_
+
+\par
+When extracting the type, programmers are encouraged to use \ref type_from to
+make it clear that the `::type` that would otherwise appear is not an
+evaluation of a lazy metafunction.  See \ref always or \ref identity for
+similar functionality that is intended for use as a metafunction.
```

Comment:
Do you mean

"When extracting the type, programmers are encouraged to use `tmpl::type_from<tmpl::type_<T>>` instead of `tmpl::type_<T>::type` because the latter looks like the evaluation of a lazy metafunction, which might cause confusion." ?

Also, is one supposed to prefer `tmpl::always` or `tmpl::identity`?

Reply:
Yes.

Brigand classifies `identity` as a mathematical operator, so I guess you could go with that as a rule.  From a practical standpoint, `always`, `identity`, and `type_from` all behave identically.